### PR TITLE
OpenTelemetry Tracing for Currency Service

### LINKED ISSUE #311: Add Tracing to Currency Service
...and use OpenTelemetry for it!

Subissue of #132. Currency Service currently does not have any telemetry instrumentation, so this is not a "conversion" task but I'm still grouping it under the same project.

### PR DESCRIPTION
**Pull request by gzhao408**
Resolves #311, subtask of #132 

Currency service acts as a server to Frontend (GetSupportedCurrencies() and Convert()) and the Checkout service in PlaceOrder (Convert()).

To manually test this branch, you can see it from (1) home page or (2) at checkout.

Frontend uses `GetSuportedCurrencies` to show the menu of currencies; the corresponding trace should show both `client-hipstershop.currencyService` (which comes from frontend) and `grpc.hipstershop.CurrencyService` (which comes from CurrencyService itself and means the context was properly propagated through grpc). 
AND It should also show the HTTP call to the ECB website 💶  🤑  💸  💵 
image

For Checkout, you need to add an item to the cart and click check out. Checkout.PlaceOrder() dispatches 6 calls to the other services, one of which is the CurrencyService to Convert the total price. The trace should show all 6 calls from the client (checkout service)`Span.client-hipstershop`. This should already show on current master:
image

On this branch, there should also be additional information from the Currency service itself. Clicking the carrot and checking Show Events should show more detail, indicating that the CurrencyService indeed receives the trace context and starts a child span `grpc.hipstershop.CurrencyService/Convert`:
image
image

### CODE REVIEW

**Comment by gzhao408** [comment]
discovered an auth issue, making a draft while i figure it out

File: src/currencyservice/server.js [hunk 1]
```diff
@@ -116,16 +138,28 @@ function _carry (amount) {
  * Lists the supported currencies
  */
 function getSupportedCurrencies (call, callback) {
+  const currentSpan = tracer.getCurrentSpan();
+  const span = tracer.startSpan('currencyservice:GetSupportedCurrencies()', {
+    parent: currentSpan,
+    kind: 1, //server
```
Comment (simonz130) [inline src/currencyservice/server.js:144]: why it's server?

File: src/currencyservice/server.js [hunk 2]
```diff
@@ -116,16 +138,28 @@ function _carry (amount) {
...
 }
 
 /**
  * Converts between currencies
  */
 function convert (call, callback) {
+  const currentSpan = tracer.getCurrentSpan();
+  const span = tracer.startSpan('currencyservice:Convert()', {
+    parent: currentSpan,
+    kind: 1, //server
```
Comment (simonz130) [inline src/currencyservice/server.js:161]: ditto

File: src/currencyservice/server.js (hunk 1 above)
Comment (gzhao408) [reply src/currencyservice/server.js:144]: Currency Service is the server and whoever is calling this function (i.e. frontend or checkout) is the client

**Comment by Daniel-Sanche** [comment]
It looks like the Ci is failing because emailservice and recommendationservice are crashing without GCP credentials. I'm not sure why that would be impacting this PR and not master though...

**Comment by gzhao408** [comment]
Rebased after #320 wooooo

**Comment by gzhao408** [comment]
I removed the @google-cloud/trace-agent because it was messing with the opentelemetry tracing, I think the exporter gets confused about which traces it's supposed to use, or one overrides the other or something. It was also the source of the "PROJECT_ID not found" warning. We don't need it 🤷

File: src/currencyservice/package.json [hunk 3]
```diff
@@ -10,9 +10,14 @@
   "license": "Apache-2.0",
   "dependencies": {
     "@google-cloud/debug-agent": "3.2.0",
+    "@google-cloud/opentelemetry-cloud-trace-exporter": "^0.4.1",
     "@google-cloud/profiler": "4.0.2",
-    "@google-cloud/trace-agent": "2.11.0",
+    "@google-cloud/trace-agent": "5.1.0",
```
Comment (simonz130) [inline src/currencyservice/package.json:15]: is this needed now that we use opentelemetry tracing?

Comment (gzhao408) [reply src/currencyservice/package.json:15]: :'((((( yeah i tried to delete it... and suddenly thousands of `grpc.google.devtools.cloudtrace.v2.TraceService/BatchWriteSpans` traces were coming out of nowhere, which caused the quota exceeded and probably the CI error. still trying to figure out what's going on, so i'm looking to see what variables i can turn on/off

**Comment by gzhao408** [comment]
Sorry for the many many force pushes. I think I've resolved the spammy batchwrite issue by switching out the processor and flushing the exporter, in doing so, accidentally enabled a new feature - the traces are so amazing that we can now track the calls to the ECB website:

image

Also I got rid of @google-cloud/trace-agent.

File: src/currencyservice/package.json (hunk 3 above)
Comment (gzhao408) [reply src/currencyservice/package.json]: alright it's GONE

**Pull request merged**